src/pulling_gauge.rs

```diff
@@ -0,0 +1,75 @@
+use std::{collections::HashMap, fmt, sync::Arc};
+
+use crate::{
+    core::Collector,
+    proto::{Gauge, Metric, MetricFamily, MetricType},
+};
+use protobuf::RepeatedField;
+
+/// A prometheus gauge that reads the value from a provided function on every
+/// collect run. Used to expose metrics that are better to be polled as opposed
+/// to continuously set.
+#[derive(Clone)]
+pub struct PullingGauge {
+    desc: crate::core::Desc,
+    value: Arc<Box<dyn Fn() -> f64 + Send + Sync>>,
+}
+
+impl fmt::Debug for PullingGauge {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("PullingGauge")
+            .field("desc", &self.desc)
+            .field("value", &"<opaque>")
+            .finish()
+    }
+}
+
+impl PullingGauge {
+    /// Create a new PullingGauge.
+    pub fn new<S1: Into<String>, S2: Into<String>>(
+        fq_name: S1,
+        help: S2,
+        value: Box<dyn Fn() -> f64 + Send + Sync>,
+    ) -> crate::Result<Self> {
+        Ok(PullingGauge {
+            value: Arc::new(value),
+            desc: crate::core::Desc::new(fq_name.into(), help.into(), Vec::new(), HashMap::new())?,
+        })
+    }
+
+    /// Create a new PullingGauge and register it with the global registry.
+    pub fn create_and_register<S1: Into<String>, S2: Into<String>>(
```

Comment:
Unless you have a strong need for this, I'd rather drop this and leave the registration logic to library consumers.
The rationale is that the global default registry is an unfortunate thing, and I'd rather not push more users to that.

Reply:
We generally are happy users of the global registry (we mostly write gRPC services using Rust for which it is convenient to have crates register standardized metrics _magically_) but I haven't actually found a place where we use this method in our internal code -- happy to remove it.